code/datums/components/infective.dm

```diff
@@ -55,19 +55,31 @@
 	if(is_weak && !prob(weak_infection_chance))
 		return
 
-	for(var/V in diseases)
-		eater.ForceContractDisease(V)
+	for(var/datum/disease/disease in diseases)
+		if(!disease.has_required_infectious_organ(eater, ORGAN_SLOT_STOMACH))
+			continue
+
+		eater.ForceContractDisease(disease)
+
 	try_infect(feeder, BODY_ZONE_L_ARM)
 
 /datum/component/infective/proc/try_infect_drink(datum/source, mob/living/drinker, mob/living/feeder)
 	SIGNAL_HANDLER
 
-	for(var/disease in diseases)
-		drinker.ForceContractDisease(disease)
 	var/appendage_zone = feeder.held_items.Find(source)
 	appendage_zone = appendage_zone == 0 ? BODY_ZONE_CHEST : appendage_zone % 2 ? BODY_ZONE_R_ARM : BODY_ZONE_L_ARM
 	try_infect(feeder, appendage_zone)
```

Comment:
So this existed before my changes. I believe the intent is that it tries to infect the person holding the object since they are making contact with the target. (kind of like a hug)  It runs a physical contact infection attempt since they are holding the item.